Test failures with openssl configured with no-ssl3-method

I am getting the following three test failures when running with openssl configured with no-ssl3-method:

```
============================= test session starts ==============================
platform linux -- Python 3.5.1, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
OpenSSL: b'OpenSSL 1.0.2g  1 Mar 2016'
cryptography: 1.3.1
rootdir: /build/pyopenssl/src/pyOpenSSL-16.0.0, inifile: setup.cfg
collected 466 items

tests/test_crypto.py .............................................................................................................................................................................................................................................
tests/test_rand.py .................
tests/test_ssl.py ..................................Fs.....s.....s..s..........s.s......s.......s...s......................................................F.s...........s................s...................................F..s..
tests/test_tsafe.py .
tests/test_util.py .

=================================== FAILURES ===================================
___________________________ ContextTests.test_method ___________________________

self = <tests.test_ssl.ContextTests testMethod=test_method>

    def test_method(self):
        """
            :py:obj:`Context` can be instantiated with one of
            :py:obj:`SSLv2_METHOD`, :py:obj:`SSLv3_METHOD`,
            :py:obj:`SSLv23_METHOD`, :py:obj:`TLSv1_METHOD`,
            :py:obj:`TLSv1_1_METHOD`, or :py:obj:`TLSv1_2_METHOD`.
            """
        methods = [
            SSLv3_METHOD, SSLv23_METHOD, TLSv1_METHOD]
        for meth in methods:
>           Context(meth)

tests/test_ssl.py:516:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <OpenSSL.SSL.Context object at 0x7f272329e0b8>, method = 2

    def __init__(self, method):
        """
            :param method: One of SSLv2_METHOD, SSLv3_METHOD, SSLv23_METHOD, or
                TLSv1_METHOD.
            """
        if not isinstance(method, integer_types):
            raise TypeError("method must be an integer")

        try:
            method_func = self._methods[method]
        except KeyError:
>           raise ValueError("No such protocol")
E           ValueError: No such protocol

src/OpenSSL/SSL.py:472: ValueError
________________ ConnectionTests.test_set_session_wrong_method _________________

self = <tests.test_ssl.ConnectionTests testMethod=test_set_session_wrong_method>

    def test_set_session_wrong_method(self):
        """
            If :py:obj:`Connection.set_session` is passed a :py:class:`Session`
            instance associated with a context using a different SSL method than
            the :py:obj:`Connection` is using, a :py:class:`OpenSSL.SSL.Error` is
            raised.
            """
        key = load_privatekey(FILETYPE_PEM, server_key_pem)
        cert = load_certificate(FILETYPE_PEM, server_cert_pem)
        ctx = Context(TLSv1_METHOD)
        ctx.use_privatekey(key)
        ctx.use_certificate(cert)
        ctx.set_session_id("unity-test")

        def makeServer(socket):
            server = Connection(ctx, socket)
            server.set_accept_state()
            return server

        originalServer, originalClient = self._loopback(
            serverFactory=makeServer)
        originalSession = originalClient.get_session()

        def makeClient(socket):
            # Intentionally use a different, incompatible method here.
            client = Connection(Context(SSLv3_METHOD), socket)
            client.set_connect_state()
            client.set_session(originalSession)
            return client

        self.assertRaises(
            Error,
>           self._loopback, clientFactory=makeClient, serverFactory=makeServer)

tests/test_ssl.py:2681:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/util.py:283: in failUnlessRaises
    f(*args, **kwargs)
tests/test_ssl.py:262: in _loopback
    client = clientFactory(client)
tests/test_ssl.py:2674: in makeClient
    client = Connection(Context(SSLv3_METHOD), socket)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <OpenSSL.SSL.Context object at 0x7f272329eac8>, method = 2

    def __init__(self, method):
        """
            :param method: One of SSLv2_METHOD, SSLv3_METHOD, SSLv23_METHOD, or
                TLSv1_METHOD.
            """
        if not isinstance(method, integer_types):
            raise TypeError("method must be an integer")

        try:
            method_func = self._methods[method]
        except KeyError:
>           raise ValueError("No such protocol")
E           ValueError: No such protocol

src/OpenSSL/SSL.py:472: ValueError
__________________ MemoryBIOTests.test_socketOverridesMemory ___________________

self = <tests.test_ssl.MemoryBIOTests testMethod=test_socketOverridesMemory>

    def test_socketOverridesMemory(self):
        """
            Test that :py:obj:`OpenSSL.SSL.bio_read` and
            :py:obj:`OpenSSL.SSL.bio_write` don't work on
            :py:obj:`OpenSSL.SSL.Connection`() that use sockets.
            """
>       context = Context(SSLv3_METHOD)

tests/test_ssl.py:3529:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <OpenSSL.SSL.Context object at 0x7f2720f15eb8>, method = 2

    def __init__(self, method):
        """
            :param method: One of SSLv2_METHOD, SSLv3_METHOD, SSLv23_METHOD, or
                TLSv1_METHOD.
            """
        if not isinstance(method, integer_types):
            raise TypeError("method must be an integer")

        try:
            method_func = self._methods[method]
        except KeyError:
>           raise ValueError("No such protocol")
E           ValueError: No such protocol

src/OpenSSL/SSL.py:472: ValueError
=============== 3 failed, 450 passed, 13 skipped in 5.15 seconds ===============
```

It would be nice to skip the tests when the method is not present.

Still failing in 16.2.0 release:

```
============================= test session starts ==============================
platform linux -- Python 3.5.2, pytest-3.0.3, py-1.4.31, pluggy-0.4.0
OpenSSL: b'OpenSSL 1.0.2j  26 Sep 2016'
cryptography: 1.5.2
rootdir: /build/pyopenssl/src/pyOpenSSL-16.2.0, inifile: setup.cfg
collected 482 items

tests/test_crypto.py ...........................................................................................................................................................................................................................................................
tests/test_rand.py .................
tests/test_ssl.py ...................................s.....s.....s..s..........s.s......s.......s...s......................................................F.s...........s................s......................................s....
tests/test_tsafe.py .
tests/test_util.py .

=================================== FAILURES ===================================
________________ ConnectionTests.test_set_session_wrong_method _________________

self = <OpenSSL.SSL.Context object at 0x7fe1befd3240>, method = 2

    def __init__(self, method):
        """
            :param method: One of SSLv2_METHOD, SSLv3_METHOD, SSLv23_METHOD, or
                TLSv1_METHOD.
            """
        if not isinstance(method, integer_types):
            raise TypeError("method must be an integer")

        try:
>           method_func = self._methods[method]
E           KeyError: 2

src/OpenSSL/SSL.py:466: KeyError

During handling of the above exception, another exception occurred:

self = <tests.test_ssl.ConnectionTests testMethod=test_set_session_wrong_method>

    def test_set_session_wrong_method(self):
        """
            If :py:obj:`Connection.set_session` is passed a :py:class:`Session`
            instance associated with a context using a different SSL method than
            the :py:obj:`Connection` is using, a :py:class:`OpenSSL.SSL.Error` is
            raised.
            """
        # Make this work on both OpenSSL 1.0.0, which doesn't support TLSv1.2
        # and also on OpenSSL 1.1.0 which doesn't support SSLv3. (SSL_ST_INIT
        # is a way to check for 1.1.0)
        if SSL_ST_INIT is not None:
            v1 = TLSv1_METHOD
            v2 = SSLv3_METHOD
        else:
            v1 = TLSv1_2_METHOD
            v2 = TLSv1_METHOD

        key = load_privatekey(FILETYPE_PEM, server_key_pem)
        cert = load_certificate(FILETYPE_PEM, server_cert_pem)
        ctx = Context(v1)
        ctx.use_privatekey(key)
        ctx.use_certificate(cert)
        ctx.set_session_id("unity-test")

        def makeServer(socket):
            server = Connection(ctx, socket)
            server.set_accept_state()
            return server

        def makeOriginalClient(socket):
            client = Connection(Context(v1), socket)
            client.set_connect_state()
            return client

        originalServer, originalClient = self._loopback(
            serverFactory=makeServer, clientFactory=makeOriginalClient)
        originalSession = originalClient.get_session()

        def makeClient(socket):
            # Intentionally use a different, incompatible method here.
            client = Connection(Context(v2), socket)
            client.set_connect_state()
            client.set_session(originalSession)
            return client

        self.assertRaises(
            Error,
>           self._loopback, clientFactory=makeClient, serverFactory=makeServer)

tests/test_ssl.py:2685:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/util.py:283: in failUnlessRaises
    f(*args, **kwargs)
tests/test_ssl.py:251: in _loopback
    client = clientFactory(client)
tests/test_ssl.py:2678: in makeClient
    client = Connection(Context(v2), socket)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <OpenSSL.SSL.Context object at 0x7fe1befd3240>, method = 2

    def __init__(self, method):
        """
            :param method: One of SSLv2_METHOD, SSLv3_METHOD, SSLv23_METHOD, or
                TLSv1_METHOD.
            """
        if not isinstance(method, integer_types):
            raise TypeError("method must be an integer")

        try:
            method_func = self._methods[method]
        except KeyError:
>           raise ValueError("No such protocol")
E           ValueError: No such protocol

src/OpenSSL/SSL.py:468: ValueError
=============== 1 failed, 468 passed, 13 skipped in 7.56 seconds ===============
```

--- a/tests/test_ssl.py
+++ b/tests/test_ssl.py
@@ -2425,7 +2425,7 @@ class TestConnection(object):
         if SSL_ST_INIT is None:
             v1 = TLSv1_2_METHOD
             v2 = TLSv1_METHOD
-        elif hasattr(_lib, "SSLv3_method()"):
+        elif hasattr(_lib, "SSLv3_method"):
             v1 = TLSv1_METHOD
             v2 = SSLv3_METHOD
          else:

--- a/tests/test_ssl.py
+++ b/tests/test_ssl.py
@@ -2419,8 +2419,17 @@ class TestConnection(object):
         with a context using a different SSL method than the `Connection`
         is using, a `OpenSSL.SSL.Error` is raised.
         """
-        v1 = TLSv1_METHOD
-        v2 = TLSv1_2_METHOD
+        # Make this work on both OpenSSL 1.0.0, which doesn't support TLSv1.2
+        # and also on OpenSSL 1.1.0 which doesn't support SSLv3. (SSL_ST_INIT
+        # is a way to check for 1.1.0)
+        if SSL_ST_INIT is None:
+            v1 = TLSv1_2_METHOD
+            v2 = TLSv1_METHOD
+        elif hasattr(_lib, "SSLv3_method()"):
+            v1 = TLSv1_METHOD
+            v2 = SSLv3_METHOD
+         else:
+            pytest.skip("Test requires either OpenSSL 1.1.0 or SSLv3")
 
         key = load_privatekey(FILETYPE_PEM, server_key_pem)
         cert = load_certificate(FILETYPE_PEM, server_cert_pem)

--- a/tests/test_ssl.py
+++ b/tests/test_ssl.py
@@ -2419,15 +2419,8 @@ class TestConnection(object):
         with a context using a different SSL method than the `Connection`
         is using, a `OpenSSL.SSL.Error` is raised.
         """
-        # Make this work on both OpenSSL 1.0.0, which doesn't support TLSv1.2
-        # and also on OpenSSL 1.1.0 which doesn't support SSLv3. (SSL_ST_INIT
-        # is a way to check for 1.1.0)
-        if SSL_ST_INIT is not None:
-            v1 = TLSv1_METHOD
-            v2 = SSLv3_METHOD
-        else:
-            v1 = TLSv1_2_METHOD
-            v2 = TLSv1_METHOD
+        v1 = TLSv1_2_METHOD
+        v2 = TLSv1_METHOD
 
         key = load_privatekey(FILETYPE_PEM, server_key_pem)
         cert = load_certificate(FILETYPE_PEM, server_cert_pem)

--- a/tests/test_ssl.py
+++ b/tests/test_ssl.py
@@ -2419,8 +2419,8 @@ class TestConnection(object):
         with a context using a different SSL method than the `Connection`
         is using, a `OpenSSL.SSL.Error` is raised.
         """
-        v1 = TLSv1_2_METHOD
-        v2 = TLSv1_METHOD
+        v1 = TLSv1_METHOD
+        v2 = TLSv1_2_METHOD
 
         key = load_privatekey(FILETYPE_PEM, server_key_pem)
         cert = load_certificate(FILETYPE_PEM, server_cert_pem)

--- a/tests/test_ssl.py
+++ b/tests/test_ssl.py
@@ -2428,7 +2428,7 @@ class TestConnection(object):
         elif hasattr(_lib, "SSLv3_method"):
             v1 = TLSv1_METHOD
             v2 = SSLv3_METHOD
-         else:
+        else:
             pytest.skip("Test requires either OpenSSL 1.1.0 or SSLv3")
 
         key = load_privatekey(FILETYPE_PEM, server_key_pem)

